icu4c/source/common/uloc_keytype.cpp

```diff
@@ -154,18 +131,26 @@ initFromResourceBundle(UErrorCode& sts) {
         // empty value indicates that BCP key is same with the legacy key.
         const char* bcpKeyId = legacyKeyId;
         if (bcpKeyIdLen > 0) {
-            char* bcpKeyIdBuf = (char*)uprv_malloc(bcpKeyIdLen + 1);
+            icu::CharString* bcpKeyIdBuf = gKeyTypeStringPool->create();
             if (bcpKeyIdBuf == NULL) {
                 sts = U_MEMORY_ALLOCATION_ERROR;
                 break;
             }
-            u_UCharsToChars(uBcpKeyId, bcpKeyIdBuf, bcpKeyIdLen);
-            bcpKeyIdBuf[bcpKeyIdLen] = 0;
-            gKeyTypeStringPool->addElement(bcpKeyIdBuf, sts);
+            int32_t resultCapacity;
+            char* buffer = bcpKeyIdBuf->getAppendBuffer(
```

Comment:
Clunky... Please replace ures_getString() + getAppendBuffer() + u_UCharsToChars()

with ures_getUnicodeString() + bckKeyIdBuf->appendInvariantChars(uBcpKeyId, sts)

Use uBcpKeyId.isEmpty() rather than `bcpKeyIdLen > 0`